From 64e721aa94e73b258be9a5fe09fc60dbcffabfd8 Mon Sep 17 00:00:00 2001
From: Tommy Webb <tommy@calyxinstitute.org>
Date: Mon, 17 Jun 2024 17:18:23 +0000
Subject: [PATCH] Redact location exif for older or legacy apps

When strict location redaction is enabled, do not grant special
treatment to apps with a pre-T target SDK or that "require" legacy
storage access; such apps should have their location metadata redacted,
too.

Test: Manual: Try to view location in Details of LeafPic Revived v0.8.7,
and the location should be zeroed out. (It may need to be a new file or
else the app may need data cleared.) Share to Phone Saver 5.2.1, and the
saved file should have zeroed-out metadata.

Issue: calyxos#2468
Change-Id: I9e5bea0e416c1d958223102a0be9eeeddfc42bf1
---
 .../android/providers/media/LocalCallingIdentity.java |  6 ++++++
 src/com/android/providers/media/MediaProvider.java    | 11 +++++++++--
 .../android/providers/media/util/PermissionUtils.java |  6 +++++-
 3 files changed, 20 insertions(+), 3 deletions(-)

diff --git a/src/com/android/providers/media/LocalCallingIdentity.java b/src/com/android/providers/media/LocalCallingIdentity.java
index 0570f0e..43173d5 100644
--- a/src/com/android/providers/media/LocalCallingIdentity.java
+++ b/src/com/android/providers/media/LocalCallingIdentity.java
@@ -69,6 +69,7 @@ import androidx.annotation.VisibleForTesting;
 import com.android.modules.utils.build.SdkLevel;
 import com.android.providers.media.util.Logging;
 import com.android.providers.media.util.LongArray;
+import com.android.providers.media.util.StrictLocationRedactionHelper;
 import com.android.providers.media.util.UserCache;
 
 import java.io.PrintWriter;
@@ -372,6 +373,11 @@ public class LocalCallingIdentity {
 
     private boolean hasPermissionInternal(int permission) {
         boolean targetSdkIsAtLeastT = getTargetSdkVersion() > Build.VERSION_CODES.S_V2;
+        if (StrictLocationRedactionHelper.getInstance(context).isSettingEnabled()) {
+            // We lie about target SDK because lower SDK should not be rewarded with access
+            // to location.
+            targetSdkIsAtLeastT = true;
+        }
         // While we're here, enforce any broad user-level restrictions
         if ((uid == Process.SHELL_UID) && context.getSystemService(UserManager.class)
                 .hasUserRestriction(UserManager.DISALLOW_USB_FILE_TRANSFER)) {
diff --git a/src/com/android/providers/media/MediaProvider.java b/src/com/android/providers/media/MediaProvider.java
index 99e0b63..c514484 100644
--- a/src/com/android/providers/media/MediaProvider.java
+++ b/src/com/android/providers/media/MediaProvider.java
@@ -10454,9 +10454,15 @@ public class MediaProvider extends ContentProvider {
      * </ul>
      */
     private boolean shouldBypassFuseRestrictions(boolean forWrite, String filePath) {
+        return shouldBypassFuseRestrictions(forWrite, filePath, /* allowLegacy */ true);
+    }
+    private boolean shouldBypassFuseRestrictions(boolean forWrite, String filePath,
+            boolean allowLegacy) {
         boolean isRequestingLegacyStorage = forWrite ? isCallingPackageLegacyWrite()
                 : isCallingPackageLegacyRead();
-        if (isRequestingLegacyStorage) {
+        final boolean shouldAllowLegacy = StrictLocationRedactionHelper.getInstance(getContext())
+                .isSettingEnabled() ? allowLegacy : true;
+        if (allowLegacy && isRequestingLegacyStorage) {
             return true;
         }
 
@@ -10594,7 +10600,8 @@ public class MediaProvider extends ContentProvider {
                 clearLocalCallingIdentity(getCachedCallingIdentityForFuse(uid));
         try {
             if (!isRedactionNeeded()
-                    || shouldBypassFuseRestrictions(/* forWrite */ false, path)) {
+                    || shouldBypassFuseRestrictions(/* forWrite */ false, path,
+                            /* allowLegacy */ false)) {
                 return new long[0];
             }
 
diff --git a/src/com/android/providers/media/util/PermissionUtils.java b/src/com/android/providers/media/util/PermissionUtils.java
index 7a73ea8..84e2afa 100644
--- a/src/com/android/providers/media/util/PermissionUtils.java
+++ b/src/com/android/providers/media/util/PermissionUtils.java
@@ -159,11 +159,15 @@ public class PermissionUtils {
     public static boolean checkPermissionAccessMediaLocation(@NonNull Context context, int pid,
             int uid, @NonNull String packageName, @Nullable String attributionTag,
             boolean isTargetSdkAtLeastT) {
+        // We lie about target SDK because lower SDK should not be rewarded with access to location.
+        final boolean reportTargetSdkAtLeastT =
+                StrictLocationRedactionHelper.getInstance(context).isSettingEnabled()
+                        ? true : isTargetSdkAtLeastT;
         final boolean hasAccessMediaLocation = checkPermissionForDataDelivery(
                 context, ACCESS_MEDIA_LOCATION, pid, uid, packageName,
                 attributionTag, generateAppOpMessage(packageName, sOpDescription.get()))
                 || checkPermissionAccessMediaCompatGrant(context, pid, uid, packageName,
-                attributionTag, isTargetSdkAtLeastT);
+                attributionTag, reportTargetSdkAtLeastT);
         // We want to be as strict as or stricter than this. This means that even if the app can
         // manage media, that's not enough if it is missing ACCESS_MEDIA_LOCATION.
         if (!hasAccessMediaLocation) {
-- 
2.43.0

